Adds support for providing Actor type name during runtime.

bjorkstromm opened the pull request:
# Description

Adds support for providing Actor type name during runtime.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #677

Duplicate of #680 with latest fixes (https://github.com/dapr/dotnet-sdk/pull/680#discussion_r646191590) which got closed unintentionally and I cannot reopen it. //cc @rynowak 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

halspang reviewed (commented):
Looks good overall, just one question.

bjorkstromm commented on src/Dapr.Actors/Runtime/ActorTypeInformation.cs:
```diff
@@ -67,11 +67,12 @@ private ActorTypeInformation()
         /// <para>2. actorType does not implement an interface deriving from <see cref="IActor"/> and is not marked as abstract.</para>
         /// </remarks>
+        [Obsolete("Use Get(Type, string) instead. This will be removed in the future.")]
         public static bool TryGet(Type actorType, out ActorTypeInformation actorTypeInformation)
```
Sorry for the late reply. Sure we can add it, but AFAIK it wasn't used within Dapr SDK at least. It's however public, so anyone could use it.  @rynowak said in the previous PR (which got accidentially closed) to just remove it. See https://github.com/dapr/dotnet-sdk/pull/680#discussion_r646191590

I don't really see a need for `TryGet` here. Most consumers shouldn't even use this API, they'll only use the higher levele `ActorRegistrationCollection.RegisterActor<T>()` API.

bjorkstromm commented:
@halspang rebased and fixed merge conflict.

saber-wang commented:
Is there a plan for this?

garrypedersen commented:
We're currently working on a massive project that would greatly benefit from this. Is there an ETA on when this will be completed?

halspang commented:
@bjorkstromm - Sorry this fell off my radar a bit. Looks like there are new merge conflicts. If we can get those sorted I think we can get this in.

msfussell commented:
Given this is going to affect the docs, before we merge this PR, can we go over an example of using the API here in the comments, so that it is clear how this is applied and what scenarios? We would also need help with writing the docs.

bjorkstromm commented:
> Given this is going to affect the docs, before we merge this PR, can we go over an example of using the API here in the comments, so that it is clear how this is applied and what scenarios? We would also need help with writing the docs.

Sure!

This PR gives the possibility to set Actor type name at runtime when registering the actors

Example from a ASP.NET Core minimal API
```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddActors(options =>
{
    // Using the new overload, we can register actor with a custom name at startup
    options.Actors.RegisterActor<MyActor>("SomeCustomActorTypeName");
    
    // This was the old approach,
    // which either user TypeName set via ActorAttribute,
    // or just the type name if ActorAttribute is not set.
    options.Actors.RegisterActor<AnotherActor>();
});

var app = builder.Build();

app.UseRouting();
app.UseEndpoints(endpoints =>
{
    endpoints.MapActorsHandlers();
});

app.Run();
```

bjorkstromm commented:
> @bjorkstromm - Sorry this fell off my radar a bit. Looks like there are new merge conflicts. If we can get those sorted I think we can get this in.

Rebased, fixed merge conflicts and signed-off..

Can we kick integration tests for .NET Core 3.1, it seem to have failed?

rynowak commented:
Looks good to go now, right @halspang ?
